projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -0,0 +1,76 @@
+import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+
+export class GraphLayout {
+
+  private levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+  private nodeToLevelMap = new Map<RenderableTopologyNode, number>();
+  private nodeCoOrdinatesMap: Map<TopologyNode, TopologyCoordinates> = new Map<TopologyNode, TopologyCoordinates>();
+
+  public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
+
+    this.findRootNodes(topology);
+    this.fillNodeAndLevelMaps();
+    this.fillNodeCoOrdinatesMap();
+    this.setNodeCoOrdinates(topology);
+  }
+
+  private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>) {
+    topology.nodes.forEach(node => {
+      if (node.incoming.length === 0) {
+        this.nodeToLevelMap.set(node, 0);
+        this.levelToNodesMap.has(0) ? this.levelToNodesMap.get(0)?.push(node) : this.levelToNodesMap.set(0, [node])
+      }
+    });
+  }
+
+  private fillNodeAndLevelMaps() {
```

Comment:
personal pref: doing this with recursion rather than a mutating while loop would probably be more readable.

Reply:
It was just a choice between going Depth First or Breadth First and since all our logic is revolving around the levels I thought it would be better to go with the level order algorithm

Reply:
Either algo can be implemented either way.

Reply:
Oh my bad, I got you wrong 😅. Since the first thing that comes to mind (at least for me) by hearing recursion I thought you were suggesting to go that route. Yeah we can implement it recursively as well.

Reply:
Made it recursive